File: CS50-and-additional-exercises/credit-card-number-validator/credit.c
#include <stdio.h>

int digitToString(void *source, char destination[], size_t destSize, char typeOfNumber)
{
    int resultOfTransform;
    if (typeOfNumber == 'l')
    {
        resultOfTransform = snprintf(destination, destSize, "%ld", *(long *)source);
    }
    else if (typeOfNumber == 'i')
    {
        resultOfTransform = snprintf(destination, destSize, "%d", *(int *)source);
    } else {
        printf("INVALID\n");
        return 0;
    }

    if (resultOfTransform < 0 || resultOfTransform >= destSize)
    {
        printf("INVALID\n");
        return 0;
    }
    else
    {
        return resultOfTransform;
    }
    return 0;
}


int main()
{
    long cardToCheck;
    int resultOfCheck;
    char term;
    char stringCardNumber[17];
    while (1)
    {
        printf("Please input: ");
        resultOfCheck = scanf("%ld%c", &cardToCheck, &term);
        if (resultOfCheck != 2 || term != '\n')
            {
                while (getchar() != '\n')
                    ;

                continue;
            }
            break;
    }

    if ((resultOfCheck = digitToString(&cardToCheck, stringCardNumber, 17, 'l')) == 0)
    {
        printf("INVALID\n");
        return 0;
    }

    int digitsAddition;

    for (int i = resultOfCheck - 2; i >= 0; i -= 2)
    {
        int digit = (stringCardNumber[i] - '0') * 2;

        if (digit > 9)
        {
            char stringifyedDigits[3];
            digitToString(&digit, stringifyedDigits, 3, 'i');
            for (unsigned int index = 0; stringifyedDigits[index] != '\0'; index++)
            {
                digitsAddition += (stringifyedDigits[index] - '0');
            }
            continue;
        }
        digitsAddition += digit;
    }

    for (int i = resultOfCheck - 1; i >= 0; i -= 2)
    {
        int digit = stringCardNumber[i] - '0';
        digitsAddition = digitsAddition + digit;
    }

    if ((digitsAddition % 10) == 0)
    {
        if ((stringCardNumber[0] - '0' == 3) && resultOfCheck == 15)
        {
            printf("AMEX\n");
            return 0;
        }
        if (((stringCardNumber[0] - '0' == 4) && resultOfCheck == 15) || ((stringCardNumber[0] - '0' == 4) && resultOfCheck == 13))
        {
            printf("VISA\n");
            return 0;
        }
        if ((stringCardNumber[0] - '0' == 5) && resultOfCheck == 16)
        {
            printf("MASTERCARD\n");
            return 0;
        }
        printf("INVALID\n");
        return 0;
    }

    printf("INVALID\n");
    return 0;
}

File: CS50-and-additional-exercises/sentimental-credit/credit.py
def check_sum(card_number):
    i = len(card_number) - 1
    sum_of_all = 0
    is_second = False

    while i >= 0:
        digit = int(card_number[i])
        if is_second:
            digit *= 2
            if digit > 9:
                digit -= 9
        sum_of_all += digit
        is_second = not is_second
        i -= 1
    
    return (sum_of_all % 10) == 0

card_number = input("Number: ")

match len(card_number):
    case 16:
        if (int(card_number[0]) == 4 and check_sum(card_number)):
            print("VISA")
        elif ((int(card_number[:2]) in range(51, 56)) and check_sum(card_number)):
            print("MASTERCARD")
        else:
            print("INVALID")
    case 15:
        if (int(card_number[:2]) in (34, 37) and check_sum(card_number)):
            print("AMEX")
        else:
            print("INVALID")
    case 13:
        if (int(card_number[:1]) == 4 and check_sum(card_number)):
            print("VISA")
        else:
            print("INVALID")
    case _:
        print("INVALID")
